flambe/nn/transformer_sru.py

```diff
@@ -0,0 +1,307 @@
+                                                   bidirectional)
+
+        self.layers = _get_clones(encoder_layer, num_layers)
+        self._reset_parameters()
+
+
+class TransformerSRUDecoder(TransformerDecoder):
+    """A TransformerSRUDecoderwith an SRU replacing the FFN."""
+
+    def __init__(self,
+                 d_model: int = 512,
+                 nhead: int = 8,
+                 num_layers: int = 6,
+                 dim_feedforward: int = 2048,
+                 dropout: float = 0.1,
+                 **kwargs: Dict[str, Any]) -> None:
+        """Initialize the TransformerEncoder.
+
+        Parameters
+        ---------
+        d_model : int
+            the number of expected features in encoder/decoder inputs.
+            Default ``512``.
+        nhead : int, optional
+            the number of heads in the multiheadattention
+            Default ``8``.
+        num_layers : int
+            the number of sub-encoder-layers in the encoder (required).
+            Default ``6``.
+        dim_feedforward : int, optional
+            the inner feedforard dimension. Default ``2048``.
+        dropout : float, optional
+            the dropout percentage. Default ``0.1``.
+
+        Extra keyword arguments are passed to the SRUCell.
+
+        """
+        decoder_layer = TransformerSRUDecoderLayer(d_model,
+                                                   nhead,
+                                                   dim_feedforward,
+                                                   dropout)
+
+        self.layers = _get_clones(decoder_layer, num_layers)
+        self._reset_parameters()
+
+
+class TransformerSRUEncoderLayer(Module):
+    """A TransformerSRUEncoderLayer with an SRU replacing the FFN."""
+
+    def __init__(self,
+                 d_model: int,
+                 nhead: int,
+                 dim_feedforward: int = 2048,
+                 dropout: float = 0.1,
+                 bidirectional: bool = False,
+                 **kwargs: Dict[str, Any]) -> None:
+        """Initialize a TransformerSRUEncoderLayer.
+
+        Parameters
+        ----------
+        d_model : int
+            The number of expected features in the input.
+        n_head : int
+            The number of heads in the multiheadattention models.
+        dim_feedforward : int, optional
+            The dimension of the feedforward network (default=2048).
+        dropout : float, optional
+            The dropout value (default=0.1).
+        bidirectional: bool
+            Whether the SRU module should be bidrectional.
+            Defaul ``False``.
+
+        Extra keyword arguments are passed to the SRUCell.
+
+        """
+        super(TransformerSRUEncoderLayer, self).__init__()
+
+        self.self_attn = nn.MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.sru = SRUCell(d_model,
+                           dim_feedforward,
+                           dropout,
```

Comment:
oh i forgot one thing. maybe introduce a `sru_dropout` here option for compatibility with the machine translation experiments of the SRU paper. i mean sth:
```
self.sru = SRUCell(d_model,
        dim_feedforward,
        dropout if sru_dropout == -1 else sru_dropout,
        ...)
```

Reply:
Done